account_invoice_pricelist/model/account_invoice.py

```diff
@@ -30,29 +12,38 @@ class AccountInvoice(models.Model):
     pricelist_id = fields.Many2one(
         comodel_name='product.pricelist', string='Pricelist',
         help="The pricelist of the partner, when the invoice is created"
-        " or the partner has changed. This is a technical field used"
-        " for reporting.")
+             " or the partner has changed."
+    )
 
-    @api.multi
-    def onchange_partner_id(
-            self, type, partner_id, date_invoice=False, payment_term=False,
-            partner_bank_id=False, company_id=False):
-        partner_obj = self.env['res.partner']
-        res = super(AccountInvoice, self).onchange_partner_id(
-            type, partner_id, date_invoice=date_invoice,
-            payment_term=payment_term, partner_bank_id=partner_bank_id,
-            company_id=company_id)
-        pricelist_id = False
-        if partner_id:
-            partner = partner_obj.browse(partner_id)
-            if type in ('out_invoice', 'out_refund'):
+    @api.onchange('partner_id', 'company_id')
+    def _onchange_partner_id(self):
+        result = super(AccountInvoice, self)._onchange_partner_id()
+        if self.partner_id:
+            if self.type in ('out_invoice', 'out_refund'):
                 # Customer Invoices
-                pricelist_id = partner.property_product_pricelist.id
-            elif type in ('in_invoice', 'in_refund'):
+                self.pricelist_id = self.partner_id.property_product_pricelist
+            elif self.type in ('in_invoice', 'in_refund'):
                 # Supplier Invoices
-                if partner._model._columns.get(
-                        'property_product_pricelist_purchase', False):
-                    pricelist_id =\
-                        partner.property_product_pricelist_purchase.id
-        res['value']['pricelist_id'] = pricelist_id
-        return res
+                if hasattr(self.partner_id,
+                           'property_product_pricelist_purchase'):
+                    self.pricelist_id =\
+                        self.partner_id.property_product_pricelist_purchase
```

Comment:
Partner model has attribute property_product_pricelist_purchase which is defined in product module. 
Why a condition here ?